### ISSUE TITLE
Constraints differences with the paper

### ISSUE THREAD

**Issue by Arsakes** [sefibk/KernelGAN#4]
To my knowledge torch `nn.L1Loss` which you  use, performs a reduction over the axes using mean operation.
Therefore when it is applied, for instance, to calculate sparsity loss, it divides the sum by factor of 169 (size of the kernel map).

This behaviour is not in agreement with the paper, where (for given sparsity constraint) summation is usead instead of a mean.
Could you specify which way of computing constraints is correct (or perhaps i'm wrong about the behaviour of L1loss or I've missed somehting).

I've tried to reproduce your results in tensorflow with not much success, the results depend strongly on hyperparameters, initialization and the way constraints are applied in time. 
I would love to read about your experiments and what you discovered about the stability of the method. 
The paper does not provide much information in that regard.

**Comment by sefibk** [comment]
You are correct about the L1 loss. 
**The code is the full implementation details** - sorry for misleading with this detail in the paper, although note it changes only the gradient **magnitude** (i.e. a factor). 

The network is image-specific and re-trained for every image - variance exists!
Experiments show variance in the specific kernel estimation but is consistent across the dataset SR results.
